### ISSUE TITLE
serde breaks when using from_reader with SmolStr

### ISSUE THREAD

**Issue by bbqsrc** [rust-analyzer/smol_str#14]
I've made a demonstration of the failures in the form of tests here: https://github.com/bbqsrc/smol_str/commit/e9480f45b7d6e9902ac83fdcc314ce5bf8e7ad9d

All of the `_reader` tests fail with the messages:

```
---- serde_tests::test_serde_reader stdout ----
thread 'serde_tests::test_serde_reader' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: string \"Hello, World\", expected a borrowed string", line: 1, column: 14)', src/libcore/result.rs:1165:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

---- serde_tests::test_serde_hashmap_reader stdout ----
thread 'serde_tests::test_serde_hashmap_reader' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: string \"a\", expected a borrowed string", line: 1, column: 5)', src/libcore/result.rs:1165:5

---- serde_tests::test_serde_struct_reader stdout ----
thread 'serde_tests::test_serde_struct_reader' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: string \"Hello, World\", expected a borrowed string", line: 1, column: 19)', src/libcore/result.rs:1165:5

---- serde_tests::test_serde_vec_reader stdout ----
thread 'serde_tests::test_serde_vec_reader' panicked at 'called `Result::unwrap()` on an `Err` value: Error("invalid type: string \"\", expected a borrowed string", line: 1, column: 3)', src/libcore/result.rs:1165:5
```

Tried on the latest stable and the following `nightly` to confirm:
```
$ rustc --version
rustc 1.39.0-nightly (66bf391c3 2019-09-23)

**Comment by matklad** [comment]
Makes sense: for deseializaition, we try to borrow the string, which is not possible with the reader. I guess the right logic here is to try borrowing first, and, if that fails, read into a `String` and then `.into` it? Not sure how to best express this with serde

**Comment by bbqsrc** [comment]
Changing `<&'de str>` to `<String>` lets the tests pass, so definitely that. Mm.

**Comment by bbqsrc** [comment]
The `D` is consumed and can't be borrowed it seems, so I'm not sure what we can do here.

**Comment by matklad** [comment]
I'll be fine with just changing this to `String` and adding tests. However, it seems it is posisble to dynamically chose between the two versions. At least, this code for Cow looks like it does something like this?

https://github.com/serde-rs/serde/blob/629802f2abfd1a54a6072992888fea7ca5bc209f/serde/src/private/de.rs#L56-L125

**Comment by bbqsrc** [comment]
I will investigate this as I have a vested interest in this loading quickly. :smile:

**Comment by bbqsrc** [comment]
Fixed. :smile:

**Comment by matklad** [comment]
published as 0.1.13
